File: app/src/main/java/ua/fvadevand/reminderstatusbar/ui/RemindersFragment.java
package ua.fvadevand.reminderstatusbar.ui;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import ua.fvadevand.reminderstatusbar.R;
import ua.fvadevand.reminderstatusbar.adapters.ReminderAdapter;
import ua.fvadevand.reminderstatusbar.adapters.ReminderAdapter.OnReminderClickListener;
import ua.fvadevand.reminderstatusbar.listeners.FabVisibilityChangeListener;

public class RemindersFragment extends Fragment {

    public static final String TAG = "RemindersFragment";
    private FabVisibilityChangeListener mFabVisibilityChangeListener;
    private OnReminderClickListener mOnReminderClickListener;
    private ReminderAdapter mAdapter;

    public RemindersFragment() {
    }

    public static RemindersFragment newInstance() {
        return new RemindersFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_reminders, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        RemindersViewModel viewModel = ViewModelProviders.of(getActivity()).get(RemindersViewModel.class);
        setupRecyclerView(view);
        viewModel.getReminderList().observe(this, reminders -> mAdapter.setReminderList(reminders));
    }

    private void setupRecyclerView(View view) {
        RecyclerView reminderListView = view.findViewById(R.id.reminder_list);
        reminderListView.setLayoutManager(new LinearLayoutManager(getContext()));
        mAdapter = new ReminderAdapter(mOnReminderClickListener);
        reminderListView.setAdapter(mAdapter);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mFabVisibilityChangeListener = (FabVisibilityChangeListener) context;
            mOnReminderClickListener = (OnReminderClickListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString()
                    + " must implement FabVisibilityChangeListener and OnReminderClickListener");
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        mFabVisibilityChangeListener.fabVisibilityChange(true);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mFabVisibilityChangeListener = null;
        mOnReminderClickListener = null;
    }
}

File: app/src/main/java/ua/fvadevand/reminderstatusbar/ui/RemindersFragment.kt
package ua.fvadevand.reminderstatusbar.ui

import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.ConcatAdapter
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.coroutines.flow.onEach
import ua.fvadevand.lifecycledelegates.fragmentProperty
import ua.fvadevand.lifecycledelegates.observeInLifecycle
import ua.fvadevand.reminderstatusbar.R
import ua.fvadevand.reminderstatusbar.adapters.ReminderAdapter
import ua.fvadevand.reminderstatusbar.adapters.ReminderHeaderAdapter
import ua.fvadevand.reminderstatusbar.databinding.FragmentRemindersBinding
import ua.fvadevand.reminderstatusbar.decorators.DividerItemDecoration
import ua.fvadevand.reminderstatusbar.decorators.SwipeToEditOrDeleteCallback
import ua.fvadevand.reminderstatusbar.listeners.OnReminderInteractListener

class RemindersFragment : Fragment(R.layout.fragment_reminders) {

    private val fragmentProperty by fragmentProperty()
    private val viewModel: RemindersViewModel by activityViewModels()
    private val binding by fragmentProperty.bindingByView(FragmentRemindersBinding::bind)
    private var reminderAdapter: ReminderAdapter? by fragmentProperty.delegateViewLifecycle()
    private var headerAdapter: ReminderHeaderAdapter? by fragmentProperty.delegateViewLifecycle()
    private var placeholder: View? = null
    private var onReminderInteractListener: OnReminderInteractListener? by fragmentProperty.delegateFragmentLifecycle()

    companion object {
        const val TAG = "RemindersFragment"

        fun newInstance() = RemindersFragment()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        onReminderInteractListener = context as OnReminderInteractListener
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        var adapterInit = false
        setupRecyclerView()
        viewModel.getRemindersSortedFlow().onEach { reminders ->
            if (reminders.isEmpty()) {
                (placeholder ?: inflatePlaceholder())?.isVisible = true
            } else {
                placeholder?.isVisible = false
            }
            reminderAdapter?.submitReminders(reminders) {
                if (!adapterInit) collectReminderEvents()
                adapterInit = true
            }
        }.observeInLifecycle(viewLifecycleOwner)
    }

    private fun setupRecyclerView() {
        val layoutManager = LinearLayoutManager(context)
        binding.reminderList.layoutManager = layoutManager
        val adapter = ReminderAdapter(
            ContextCompat.getColor(requireContext(), R.color.colorBackgroundHighlight),
            onReminderInteractListener
        )
        val headerAdapter = ReminderHeaderAdapter(R.string.reminders_title)
        val concatAdapter = ConcatAdapter(headerAdapter, adapter)
        binding.reminderList.adapter = concatAdapter
        binding.reminderList.addItemDecoration(
            DividerItemDecoration(
                requireContext(),
                layoutManager.orientation,
                requireContext().resources.getDimensionPixelOffset(R.dimen.item_reminder_divider_offset_start),
                requireContext().resources.getDimensionPixelOffset(R.dimen.item_reminder_divider_offset_end),
                drawInFirstItem = false,
                drawInLastItem = false
            )
        )
        ItemTouchHelper(SwipeToEditOrDeleteCallback(requireContext(), adapter))
            .attachToRecyclerView(binding.reminderList)

        reminderAdapter = adapter
        this.headerAdapter = headerAdapter
    }

    private fun collectReminderEvents() {
        viewModel.reminderEvents.onEach { reminderId ->
            val reminderIndex = reminderAdapter?.getReminderIndex(reminderId)
            val headerCount = headerAdapter?.itemCount
            if (headerCount == null || reminderIndex == null || reminderIndex == -1) return@onEach
            binding.reminderList.scrollToPosition(reminderIndex + headerCount)
            reminderAdapter?.highlightItem(reminderIndex)
        }.observeInLifecycle(viewLifecycleOwner)
    }

    private fun inflatePlaceholder(): View? {
        placeholder = binding.viewStubReminders.inflate()
        return placeholder
    }

}
